crates/viewer/re_viewer/src/app_state.rs

```diff
@@ -177,467 +172,493 @@ impl AppState {
     ) {
         re_tracing::profile_function!();
 
-        let blueprint_query = self.blueprint_query_for_viewer(store_context.blueprint);
-
-        let Self {
-            app_options,
-            recording_configs,
-            blueprint_undo_state,
-            blueprint_cfg,
-            selection_panel,
-            time_panel,
-            blueprint_time_panel,
-            blueprint_tree,
-            welcome_screen,
-            datastore_ui,
-            redap_servers,
-            display_mode,
-            show_settings_ui,
-            view_states,
-            selection_state,
-            focused_item,
-        } = self;
-
         // check state early, before the UI has a chance to close these popups
         let is_any_popup_open = ui.memory(|m| m.any_popup_open());
 
-        blueprint_undo_state
-            .entry(store_context.blueprint.store_id().clone())
-            .or_default()
-            .update(ui.ctx(), store_context.blueprint);
-
-        let viewport_blueprint =
-            ViewportBlueprint::try_from_db(store_context.blueprint, &blueprint_query);
-        let viewport_ui = ViewportUi::new(viewport_blueprint);
-
-        // If the blueprint is invalid, reset it.
-        if viewport_ui.blueprint.is_invalid() {
-            re_log::warn!("Incompatible blueprint detected. Resetting to default.");
-            command_sender.send_system(re_viewer_context::SystemCommand::ClearActiveBlueprint);
+        match (self.navigation.peek(), store_context) {
+            (DisplayMode::WelcomeScreen, _) => {
+                self.welcome_screen.ui(
+                    ui,
+                    command_sender,
+                    welcome_screen_state,
+                    is_history_enabled,
+                );
+            }
+            (DisplayMode::Settings, _) => {
```

Comment:
nit: Let's add some whitespace in that match statement for clarity

```suggestion

            (DisplayMode::Settings, _) => {
```